### PR TITLE
K8petstore

### PR DESCRIPTION
**Pull request by jayunit100**
Here it is ! I've finished automating and testing this at 20 and 40 nodes, and added a throughput indicator.  

You can see some interesting behaviour happens at times, i.e. flatlining, haven't determined quite exactly why yet, could be a minor glitch in the generator , but since its stochastic, i think its related to stability of containers in certain scenarios.  

Ive ran this on up to 40 nodes to generate up to around 500,000 transactions in total.  

Would love to scale it further.

All docker code is contained here, including the redis containers themselves so things like /etc/redis conf files can be customized, and also, so that we pegs a specific version to avoid goredis api compatibility syncing issues...  w/ excpetion of the bigpetstore data generator, which is already curated in apache bigtop (documentation is in the README for that as well).

To run ?  Just run `k8petstore.sh` :)

### CODE REVIEW

**Comment by satnam6502** [comment]
@jayunit100 please can you suggest a reviewer(s)?

**Comment by jayunit100** [comment]
Hi satnam.  Possbly @brendanburns , and yourself @satnam6502 might be able to review this .   It should be very easy to test (just run `k8petstore.sh` and modify the variables at the top, which are self explanatory, as you see fit)

**Comment by jayunit100** [comment]
(fyi, the official issue tracker is https://github.com/GoogleCloudPlatform/kubernetes/issues/5260 )

**Comment by satnam6502** [comment]
Thanks for the suggestions. I'm on call today so I've asked Brendan to help out.

File: examples/k8petstore/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+If you are new to kubernetes, and you haven't run guestbook yet, 
+
+you might want to stop here and go back and run guestbook app first.  
+
+The guestbook tutorial will teach you alot about the basics of kubernetes, and we've tried not to be redundant here.
+
+## Architecture of this SOA
+
+A diagram of the overall architecture of this application can be seen in arch.dot (you can paste the contents in any graphviz viewer, including online ones such as http://sandbox.kidstrythisathome.com/erdos/.
```
Comment (brendandburns) [inline examples/k8petstore/README.md]: linkify arch.dot ?

File: examples/k8petstore/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+- It comes with visualizations for graphing whats happening in Redis transactions, along with commandline printouts of transaction throughput
+- It is hackable : you can build all images from the files is in this repository (With the exception of the data generator, which is apache bigtop).
+- It generates massive load using a semantically rich, realistic transaction simulator for petstores
+
+This application will run a web server which returns REDIS records for a petstore application.
+It is meant to simulate and test high load on kubernetes or any other docker based system.
+
+If you are new to kubernetes, and you haven't run guestbook yet, 
+
+you might want to stop here and go back and run guestbook app first.
```
Comment (brendandburns) [inline examples/k8petstore/README.md]: link to the guestbook app readme?

File: examples/k8petstore/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+Now, how to run the entire application in kubernetes? 
+
+To simplify running this application, we have a single file, k8petstore.sh, which writes out json files on to disk.  This allows us to have dynamic parameters, without needing to worry about managing multiplejson files.
+
+You might want to change it to point to your customized Go image, if you chose to modify things.  
+
+like the number of data generators (more generators will create more load on the redis master).
+
+So, to run this app in kubernetes, simply run `k8petstore.sh`.
```
Comment (brendandburns) [inline examples/k8petstore/README.md]: linkify k8petstore.sh?

File: examples/k8petstore/README.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+## Future
+
+In the future, we plan to add cassandra support.  Redis is a fabulous in memory data store, but it is not meant for truly available and resilient storage.  
+
+Thus we plan to add another tier of queueing, which empties the REDIS transactions into a cassandra store which persists.   
+
+## Questions
+
+For questions on running this app, you can ask on the google containers group.
```
Comment (brendandburns) [inline examples/k8petstore/README.md]: google-containers@googlegroups.com or #google-containers on IRC.

**Comment by brendandburns** [comment]
Cool, some minor comments on the docs, but generally LGTM.

Thanks for doing this!
--brendan

**Comment by jayunit100** [comment]
okay updated @brendanburns  !

**Comment by brendandburns** [comment]
Sorry, one more thing, can you run gofmt on:

```./examples/k8petstore/web-server/src/main/PetStoreBook.go```

Thanks!
--brendan

**Comment by jayunit100** [comment]
okay, done.  thats a neat tool.  didn't know about it

**Comment by brendandburns** [comment]
Ugh, travis is trying to build your example code:

https://travis-ci.org/GoogleCloudPlatform/kubernetes/jobs/54292082

We should either include the right libraries in our Gopath, or block travis from building this somehow.

**Comment by jayunit100** [comment]
hiya, okay, i;ll look into adding the libraries.  Im not sure where to add them so if its non obvious just ping in this thread :)... will update when im done.

**Comment by jayunit100** [comment]
okay, pushed an to GoDeps ........... we will see if it works.  thanks for noticing that !~

**Comment by jayunit100** [comment]
Well, looks like the build still fails.  I think i have a way to reproduce faithfully now .  I guess one option is to rename `src` to `_src` .... but that is only a temporary fix.

**Comment by jayunit100** [comment]
Well, I got the build working locally: 
```
[jay@rhbd kubernetes]$  go build -v ./... 2>&1 | grep examples
_/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/examples
_/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/examples/liveness/image
_/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes/examples/k8petstore/web-server/src/main
[jay@rhbd kubernetes]$ pwd
/home/jay/gopath/src/github.com/GoogleCloudPlatform/kubernetes
```
I will see if i can follow the directions in the developer docs to build it cleanly.  maybe its just working b.c of the local deps which are on my machine, (i.e. possibly i didn't add them correctly into the Godeps json file)

**Comment by jayunit100** [comment]
@brendandburns okay, i did the temporary disablement of the `src` dir by moving to `_src`.  

As a follow on , lets work on https://github.com/GoogleCloudPlatform/kubernetes/issues/5465, as part of which we can get rid of needing to do stuff like that, and bring `examples/` into the build

**Comment by jayunit100** [comment]
(bump) tl;dr this is now passing and ready to merge - i shaded the src package to `_src`, and updated the dependencies.  maybe in a future build, it will pass once travis integrates examples/ src properly.

**Comment by smarterclayton** [comment]
Squash before merge (once final sign offs are given) please.

**Comment by jayunit100** [comment]
Squashed into 2 commits (original patch in the first ; cleanup in the second).

**Comment by brendandburns** [comment]
LGTM, will merge on green.

**Comment by jlowdermilk** [comment]
lgtm, but needs rebase and squash.

**Comment by jayunit100** [comment]
okay , rebased and squashed into 3 commits (one containing the fixes which jeff pushed, one containing the cleanups which i made after the original patch, and one with the original code dump).   thanks for your help Jeff and Brendan !

**Comment by brendandburns** [comment]
merged!

**Pull request merged**